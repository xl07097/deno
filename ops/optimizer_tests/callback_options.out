#[allow(non_camel_case_types)]
///Auto-generated by `deno_ops`, i.e: `#[op]`
///
///Use `op_fallback::decl()` to get an op-declaration
///you can include in a `deno_core::Extension`.
pub struct op_fallback;
#[doc(hidden)]
impl op_fallback {
    pub const fn name() -> &'static str {
        stringify!(op_fallback)
    }
    #[allow(clippy::not_unsafe_ptr_arg_deref)]
    pub extern "C" fn v8_fn_ptr<'scope>(
        info: *const deno_core::v8::FunctionCallbackInfo,
    ) {
        let info = unsafe { &*info };
        let scope = &mut unsafe { deno_core::v8::CallbackScope::new(info) };
        let args = deno_core::v8::FunctionCallbackArguments::from_function_callback_info(
            info,
        );
        let rv = deno_core::v8::ReturnValue::from_function_callback_info(info);
        Self::v8_func(scope, args, rv);
    }
    pub const fn decl<'scope>() -> deno_core::OpDecl {
        deno_core::OpDecl {
            name: Self::name(),
            v8_fn_ptr: Self::v8_fn_ptr as _,
            enabled: true,
            fast_fn: {
                use deno_core::v8::fast_api::Type::*;
                use deno_core::v8::fast_api::CType;
                Some(
                    deno_core::v8::fast_api::FastFunction::new(
                        &[V8Value, CallbackOptions],
                        CType::Void,
                        op_fallback_fast_fn as *const ::std::ffi::c_void,
                    ),
                )
            },
            is_async: false,
            is_unstable: false,
            is_v8: false,
            force_registration: false,
            arg_count: 1usize as u8,
        }
    }
    #[inline]
    #[allow(clippy::too_many_arguments)]
    fn call(options: Option<&mut FastApiCallbackOptions>) {
        if let Some(options) = options {
            options.fallback = true;
        }
    }
    pub fn v8_func<'scope>(
        scope: &mut deno_core::v8::HandleScope<'scope>,
        args: deno_core::v8::FunctionCallbackArguments,
        mut rv: deno_core::v8::ReturnValue,
    ) {
        let ctx = unsafe {
            &*(deno_core::v8::Local::<deno_core::v8::External>::cast(args.data()).value()
                as *const deno_core::_ops::OpCtx)
        };
        let arg_0 = None;
        let result = Self::call(arg_0);
        let op_state = ::std::cell::RefCell::borrow(&*ctx.state);
        op_state.tracker.track_sync(ctx.id);
    }
}
#[allow(clippy::too_many_arguments)]
fn op_fallback_fast_fn<'scope>(
    _: deno_core::v8::Local<deno_core::v8::Object>,
    fast_api_callback_options: *mut deno_core::v8::fast_api::FastApiCallbackOptions,
) -> () {
    use deno_core::v8;
    use deno_core::_ops;
    let __opts: &mut v8::fast_api::FastApiCallbackOptions = unsafe {
        &mut *fast_api_callback_options
    };
    let result = op_fallback::call(options);
    result
}
