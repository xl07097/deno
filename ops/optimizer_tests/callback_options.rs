fn op_fallback(options: Option<&mut FastApiCallbackOptions>) {
  if let Some(options) = options {
    options.fallback = true;
  }
}
