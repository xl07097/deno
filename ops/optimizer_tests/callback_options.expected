=== Optimizer Dump ===
returns_result: false
has_ref_opstate: false
has_rc_opstate: false
has_fast_callback_option: true
needs_fast_callback_option: false
fast_result: Some(Void)
fast_parameters: [V8Value]
transforms: {}
is_async: false
fast_compatible: true
